### PR TITLE
Update repository with a pre-check role, updated meta data files and align with Ansible best practices

### PR DESCRIPTION
**Pull request by ddimatos**
This rather large pull request started out touching many files which had the deprecated `zos_ssh` connection plugin, given so many projects were being changed, I decided to take the opportunity and address many more issues and lay the foundation down for some future needs.

This pull request:

1. Removes all `zos_ssh` connection plugins from playbooks (breaking change if not removed)
2. It introduces a `changelog.yml` to track the changes since there is really no ideal way to have official releases of this repository given a collective effort; it removes changelog entries from playbooks and transfers to `changlog.yml`
3. It standardizes on extension `.yml` over a mix of both `.yml` and `.yaml`
4. It introduces directory `inventories/` with `host_vars/*.yml`,`group_vars/all.yml` and inventory.yml in an effort:
- to align with [Ansibles best practices](https://docs.ansible.com/ansible/latest/user_guide/sample_setup.html#alternative-directory-layout)
- place the vars [higher in the precedence ordering](https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#understanding-variable-precedence)
- Organize host and group vars such that `group_vars/all.yml is common to all z/OS targets, `host_vars/*.yml` is aligned to target variables such that `ansible_python_interpreter` has been removed from inventory to `host_vars` because it better aligns to a target than it does inventory. 
- Future proof to avoid where [some commands](https://docs.ansible.com/ansible/latest/user_guide/intro_inventory.html#organizing-host-and-group-variables) such as `ansible` and `ansible-console` will only look for group_vars/ and host_vars/ in the inventory directory.
5.  Introduces a [requirements check role](https://github.com/IBM/z_ansible_collections_samples/tree/update/remove-zos-ssh-add-pre-check-role/roles/requirements-check) that will ensure that if a user is not using a supported Ansible version or have the right collections (or versions) installed, it will print an error to STDOUT. This relies on new file `runtime-requirements.yml`
6. Introduce new metadata [runtime-requirements.yml](https://github.com/IBM/z_ansible_collections_samples/blob/update/remove-zos-ssh-add-pre-check-role/zos_concepts/data_sets/data_set_basics/runtime-requirements.yml) that outlines what the playbook project supports, it also lays the foundation down for some future dependency analysis tooling to be released in the future.
7. [Simplified READMEs](https://github.com/IBM/z_ansible_collections_samples/blob/update/remove-zos-ssh-add-pre-check-role/zos_concepts/data_sets/data_set_basics/README.md#configuration)  for all projects and reusable documentation put in a share folder and linked
8. Simplified configuration, one less variable to configure, now [ansible_python_interpreter](ansible_python_interpreter) can inherit from variable `PYZ` in host_vars
9. [New doc additions](https://github.com/IBM/z_ansible_collections_samples/tree/update/remove-zos-ssh-add-pre-check-role/docs/share/zos_core) and doc lessened and simpler to read. 
10. Follow [Ansibles best practices](https://docs.ansible.com/ansible/2.4/playbooks_best_practices.html#directory-layout) and include a `site.yml` for nearly all projects, at minimum any remaining projects should consider the same and call the `pre-check-role` to ensure a user who might not see the documentations requirements at least have it programmatically caught to ensure a better experience. 
11. Several bug fixes, particularly when modules were copying content to z/OS and not taking into account the record length to the target data set. 
12. Deprecated playbook `host_setup` because the artifacts (*_vars, inventory) it generates does not adhere to the latest architecture changes, it also does not leverage the updated meta data thus there always being a need to keep updating this, a future edition will be created that is more dynamic and works with this change. Its value is debatable, configuration has been reduced to 4 variables be configured by the user, given how simple it is to configure Ansible, it may not be worth introducing it again.
13. Update `.gitignore` to block archive formats
14. Address contributors section to describe what needs to be added to a pull request (STDOUT of a playbook) and the Apache license. 
15. [Fixes wrong playbook extension in doc issue 44](https://github.com/IBM/z_ansible_collections_samples/issues/44)
16. [Fixes missing password-less doc git issue 98](https://github.com/IBM/z_ansible_collections_samples/issues/98)

### CODE REVIEW

File: zos_concepts/user_management/add_remove_user/README.md [hunk 1]
```diff
@@ -1,18 +1,23 @@
...
 names, authority, and the artifacts that will be created and cleaned up.
-Although samples are written to operate without the need for the user’s
-configuration, flexibility is written into the samples because it is not easy
-to determine if a sample has access to the host’s resources. Review the
-playbook notes sections for additional details and configuration.
 
 ## Playbook Summary
 
-- [**add-user.yml**](add-user.yml) - Handles adding a user to z/OS. Playbook includes: group configuration, granting permissions, generating passwords, creating and mounting ZFS filesystem for OMVS, transferring files and templates, creating generic profile, and creating catalog alias. Playbook is designed to be used standalone or in an Ansible Tower workflow template.
+- [**add-user.yml**](add-user.yml) - Handles adding a user to z/OS. Playbook includes; group configuration, granting permissions, generating passwords, creating and mounting ZFS filesystem for OMVS, transferring files and templates, creating generic profile, and creating catalog alias. Playbook is designed to be used standalone or in an Ansible Tower workflow template.
```
Comment (itsBryantP) [inline zos_concepts/user_management/add_remove_user/README.md:20]: remove semicolon altogether. "Playbook includes group configuration,..."

File: zos_concepts/user_management/add_remove_user/site.yml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+# © Copyright IBM Corporation 2021
+###############################################################################
+
+###############################################################################
+# This playbook (`site.yml`) is the master playbook for this project. Based on
+# the playbooks requirments, `site.yml` playbook will evalute the requirements
+# and report when the minimum requirements are not configured and fail playbook
+# execution.
+#
+# Upon successful dependency analyais, the z/OS playbook will be executed.
```
Comment (itsBryantP) [inline zos_concepts/user_management/add_remove_user/site.yml:11]: minor typo in comment throughout several site.yml files. 
"analyais"

**Review by itsBryantP — APPROVED** [review]
Review of PR complete. PR is approved under the assumption that comments will be reviewed and addressed as the author @ddimatos sees fit.

File: README.md [hunk 3]
```diff
@@ -22,7 +21,6 @@ the repository and configure your GitHub
 to receive ongoing updates about specific activity on GitHub.
```
[the remark this answers is not in the document]
Comment (ddimatos) [reply README.md:21]: Done

File: docs/share/zos_core/configuration_guide.md [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+
+The value for the property **ansible_host** is the hostname of the
+managed node; for example, `ansible_host: zvm1.vmec.svl.ibm.com`
+
+The value for the property **ansible_host** is the z/OS managed user to
+connect and run as over SSH; for example, `ansible_user: zosadm`.
+
+Behavioral inventory parameters such as `ansible_port` allow you
+to set the port for a host can be viewed in the behavioral inventory parameters
+[documentation](https://docs.ansible.com/ansible/latest/user_guide/intro_inventory.html#connecting-to-hosts-behavioral-inventory-parameters).
```
[the remark this answers is not in the document]
Comment (ddimatos) [reply docs/share/zos_core/configuration_guide.md:75]: Done

File: docs/share/zos_core/configuration_guide.md [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+Host variables are environment variables referred and located in a `host_vars/`
+directory under the `inventories/` directory. They don't necessarily have to be
+located under `inventories/`, but it makes for a more universal experience when
+using tools such as [ansible](https://docs.ansible.com/ansible/latest/cli/ansible.html),
+[ansible-console](https://docs.ansible.com/ansible/latest/cli/ansible-console.html)
+will only look for `group_vars/` and `host_vars/` in the inventory directory.
+
+`host_vars/` provides organization such that they are unique to a the z/OS
+managed node. The playbook projects include a `*.yml` usually named
+`zos_host.yaml` that is located in the `host_vars/` directory. These variables
```
[the remark this answers is not in the document]
Comment (ddimatos) [reply docs/share/zos_core/configuration_guide.md]: done

File: docs/share/zos_core/configuration_guide.md [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+path to the playbook when using the command line command `ansible-playbook`.
+
+When running a playbook, review the comments on how the playbook will
+interact with your target, ensure you have the required authority and
+permissions such as writing the the target directories or creating data sets.
+
+### Ansible command
+Use the Ansible command `ansible-playbook` to run the sample playbook.
+The command syntax is `ansible-playbook -i <inventory> <playbook>`; for
+example, `ansible-playbook -i inventories sample.yaml`.
```
[the remark this answers is not in the document]
Comment (ddimatos) [reply docs/share/zos_core/configuration_guide.md]: done

File: docs/share/zos_core/configuration_guide.md [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+chmod 700 ~/.ssh
+touch ~/.ssh/authorized_keys
+chmod 600 ~/.ssh/authorized_keys
+```
+finally, append the contents from `cat ~/.ssh/id_rsa.pub` to `~/.ssh/authorized_keys`
+
+### Password prompt
+Alternatively, you can use the `--ask-pass` option to be prompted for
+the user\'s password each time a playbook is run; for example,
+`ansible-playbook -i inventory zos-collection-sample.yaml --ask-pass`.
```
[the remark this answers is not in the document]
Comment (ddimatos) [reply docs/share/zos_core/configuration_guide.md]: Done

File: zos_basics/constructs/README.md [hunk 8]
```diff
@@ -79,29 +67,30 @@ Usage:
 
 ```
   1) Run in development mode using JSON cached input
-        ansible-playbook -i inventory zos_job_submit_ansible_constructs.yaml --tags "mode_development_json"
+        ansible-playbook -i inventory.yml zos_job_submit_ansible_constructs.yaml --tags "mode_development_json"
```
[the remark this answers is not in the document]
Comment (ddimatos) [reply zos_basics/constructs/README.md]: Done

File: zos_basics/constructs/zos_job_submit_ansible_constructs.yml [hunk 9]
```diff
@@ -39,25 +39,25 @@
 #  ansible-playbook -i <inventory> <playbook> --tags "....."
 #
 #   1) Run in development mode using JSON cached input
-#         ansible-playbook -i inventory zos_job_submit_ansible_constructs.yaml --tags "mode_development_json"
+#         ansible-playbook -i inventories zos_job_submit_ansible_constructs.yaml --tags "mode_development_json"
```
[the remark this answers is not in the document]
Comment (ddimatos) [reply zos_basics/constructs/zos_job_submit_ansible_constructs.yml]: Done

File: roles/requirements-check/tasks/collection_check.yml [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+# playbook requirements.
+################################################################################
+
+---
+- name: Extracting Ansible enviroment variable COLLECTIONS_PATHS from localhost
+  shell:
+    cmd: ansible-config dump|grep -i COLLECTIONS_PATHS|awk -F" = " '{print $2}'
+  register: ansible_collections_paths
+
+- name: DEBUG(-v) Display Ansbile COLLECTIONS_PATHS environment variable
```
[the remark this answers is not in the document]
Comment (ddimatos) [reply roles/requirements-check/tasks/collection_check.yml]: Done

File: roles/requirements-check/tasks/collection_check.yml [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+    collection_paths: "{{ ansible_collections_paths.stdout if ansible_collections_paths.stdout != \"\" else ['~/.ansible/collections'] }}"
+
+- name: Evaluate Ansible COLLECTIONS_PATHS existance and status
+  stat:
+    path: "{{ item }}"
+  register: collection_paths_status
+  with_items: "{{ collection_paths }}"
+  # ignore_errors: True (Should we?)
+
+- name: DEBUG(-v) Display Ansbile COLLECTIONS_PATHS existance and status
```
[the remark this answers is not in the document]
Comment (ddimatos) [reply roles/requirements-check/tasks/collection_check.yml]: Done

File: zos_concepts/data_sets/data_set_basics/data_set_basics.yml [hunk 12]
```diff
@@ -137,9 +117,10 @@
       debug:
         msg: "{{ result }}"
 
-    - name: Populate USS file
-      shell: "echo {{ lookup('file', playbook_dir + '/files/HELLO.jcl') | quote }} > {{ tgt_tmp_dir }}/HELLO"
-      changed_when: true
+    - zos_copy:
+        src: /Users/ddimatos/git/github/z_ansible_collections_samples/zos_concepts/data_sets/data_set_basics/files/HELLO.jcl
```
[the remark this answers is not in the document]
Comment (ddimatos) [reply zos_concepts/data_sets/data_set_basics/data_set_basics.yml]: Done

File: zos_concepts/data_sets/data_set_basics/data_set_basics.yml [hunk 13]
```diff
@@ -118,15 +96,17 @@
         name: "{{ data_set_name }}"
         type: seq
         state: present
+        replace: true
         format: fb
-        record_length: 80
+        record_length: 100
         size: 5M
+        volume: dimato
```
[the remark this answers is not in the document]
Comment (ddimatos) [reply zos_concepts/data_sets/data_set_basics/data_set_basics.yml:103]: Done

File: zos_concepts/user_management/add_remove_user/README.md (hunk 1 above)
Comment (ddimatos) [reply zos_concepts/user_management/add_remove_user/README.md]: Done

File: zos_concepts/data_sets/data_set_basics/data_set_basics.yml (hunk 12 above)
[the remark this answers is not in the document]
Comment (ddimatos) [reply zos_concepts/data_sets/data_set_basics/data_set_basics.yml]: Done

File: zos_concepts/user_management/add_remove_user/site.yml (hunk 2 above)
Comment (ddimatos) [reply zos_concepts/user_management/add_remove_user/site.yml]: Done

Comment (ddimatos) [reply zos_concepts/user_management/add_remove_user/site.yml]: Done

**Pull request merged**